src/librustc_mir/transform/const_prop.rs

```diff
@@ -553,6 +553,40 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                 }
             }
 
+            Rvalue::Cast(CastKind::Misc, op, ty) => {
+                trace!("checking Cast(Misc, {:?}, {:?})", op, ty);
+
+                if ty.is_integral() && op.ty(&self.local_decls, self.tcx).is_integral() {
+                    let value = self.use_ecx(source_info, |this| {
+                        this.ecx.read_immediate(this.ecx.eval_operand(op, None)?)
+                    })?;
+
+                    // Do not try to read bits for ZSTs
+                    if !value.layout.is_zst() {
+                        let dest_bits = place_layout.size.bits();
+                        let value_size = value.layout.size;
+                        let value_bits = value.to_scalar().and_then(|r| r.to_bits(value_size));
+                        let max_size = 1u128.checked_shl(dest_bits as u32);
+                        if let (Ok(value_bits), Some(max_size)) = (value_bits, max_size) {
+                            if value_bits >= max_size {
+                                let scope = source_info.scope;
+                                let lint_root = match &self.source_scopes[scope].local_data {
+                                    ClearCrossCrate::Set(data) => data.lint_root,
+                                    ClearCrossCrate::Clear => return None,
+                                };
+                                self.tcx.lint_hir(
+                                    ::rustc::lint::builtin::CONST_ERR,
+                                    lint_root,
+                                    span,
+                                    &format!("overflowing cast"),
```

Comment:
I think "lossy" or "truncating" is clearer than "overflowing"

Also this could emit some information like the value before and after truncation and the reason (target type is only X bits). Not sure what level of detail we want here

cc @rust-lang/wg-diagnostics

Reply:
How about `truncating cast: the value {} requires {} bits but the target type is only {}`